devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/EffectiveConfig.java

```diff
@@ -0,0 +1,219 @@
+package io.quarkus.gradle.tasks;
+
+import static java.util.Collections.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.eclipse.microprofile.config.spi.ConfigSourceProvider;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import io.quarkus.deployment.configuration.ClassLoadingConfig;
+import io.quarkus.deployment.pkg.PackageConfig;
+import io.quarkus.runtime.configuration.ConfigUtils;
+import io.smallrye.config.AbstractLocationConfigSourceLoader;
+import io.smallrye.config.EnvConfigSource;
+import io.smallrye.config.PropertiesConfigSource;
+import io.smallrye.config.SmallRyeConfig;
+import io.smallrye.config.common.utils.ConfigSourceUtil;
+import io.smallrye.config.source.yaml.YamlConfigSource;
+
+/**
+ * Helper that bundles the various sources of config options for the Gradle plugin: system environment, system properties,
+ * quarkus build properties (on the Quarkus extension), Gradle project properties, application properties and "forced/task"
+ * properties (on the Gradle task).
+ *
+ * <p>
+ * Eventually used to construct a map with the <em>effective</em> config options from all the sources above and expose
+ * the Quarkus config objects like {@link PackageConfig}, {@link ClassLoadingConfig} and the underlying {@link SmallRyeConfig}.
+ */
+final class EffectiveConfig {
+    private final Map<String, String> fullConfig;
+
+    // URLs of all application.properties/yaml/yml files that were consulted (including those that do not exist)
+    private final List<URL> applicationPropsSources;
+    private final SmallRyeConfig config;
+
+    private EffectiveConfig(Builder builder) {
+        List<ConfigSource> configSources = new ArrayList<>();
+        // TODO add io.quarkus.runtime.configuration.DefaultsConfigSource ?
+        // TODO leverage io.quarkus.runtime.configuration.ProfileManager ?
+
+        // Effective "ordinals" for the config sources:
+        // 800 -> forcedProperties
+        // 700 -> taskProperties
+        // 600 -> System.getProperties()
+        // 500 -> System.getenv()
+        // 480 -> quarkusBuildProperties
+        // 300 -> projectProperties
+        // 0-250 -> application.(properties|yaml|yml)
```

Comment:
@radcortez could you please review the config composition? Thanks.

Reply:
Didn't knew that list.
Updated the ordinals + added microprofile.properties as well.